From 3bf0d3316e83484353fa65982e602f3b13f2a7a7 Mon Sep 17 00:00:00 2001
From: Rob Clark <rclark@redhat.com>
Date: Wed, 17 Dec 2014 21:38:37 -0500
Subject: [drm] i915: Fix EIO/wedged handling in gem fault handler

Message-id: <1418852380-13061-27-git-send-email-rclark@redhat.com>
Patchwork-id: 102164
O-Subject: [RHEL7 drm 26/89] drm/i915: Fix EIO/wedged handling in gem fault handler
Bugzilla: 1153301
RH-Acked-by: Jerome Glisse <jglisse@redhat.com>
RH-Acked-by: Dave Airlie <airlied@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1153301
Upstream: since v3.16.4

commit 6f24f542e98850ce4f5d367f6797ec4999aefddc

Author:     Daniel Vetter <daniel.vetter@ffwll.ch>
AuthorDate: Thu Sep 4 09:36:18 2014 +0200
Commit:     Rob Clark <rclark@redhat.com>
CommitDate: Wed Dec 10 17:15:01 2014 -0500

    drm/i915: Fix EIO/wedged handling in gem fault handler

    commit 2232f0315c6688f5ff6b2067ea88d97542034873 upstream.

    In

    commit 1f83fee08d625f8d0130f9fe5ef7b17c2e022f3c
    Author: Daniel Vetter <daniel.vetter@ffwll.ch>
    Date:   Thu Nov 15 17:17:22 2012 +0100

        drm/i915: clear up wedged transitions

    I've accidentally inverted the EIO/wedged handling in the fault
    handler: We want to return the EIO as a SIGBUS only if it's not
    because of the gpu having died, to prevent userspace from unduly
    dying.

    In my defence the comment right above is completely misleading, so fix
    both.

    v2: Drop the WARN_ON, it's not actually a bug to e.g. receive an -EIO
    when swap-in fails.

    v3: Don't remove too much ... oops.

    Reported-by: Chris Wilson <chris@chris-wilson.co.uk>
    Cc: Chris Wilson <chris@chris-wilson.co.uk>
    Signed-off-by: Daniel Vetter <daniel.vetter@intel.com>
    Reviewed-by: Chris Wilson <chris@chris-wilson.co.uk>
    Signed-off-by: Jani Nikula <jani.nikula@intel.com>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

Signed-off-by: Rob Clark <rclark@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/gpu/drm/i915/i915_gem.c b/drivers/gpu/drm/i915/i915_gem.c
index a5b077c..53b4db8 100644
--- a/drivers/gpu/drm/i915/i915_gem.c
+++ b/drivers/gpu/drm/i915/i915_gem.c
@@ -1575,10 +1575,13 @@ unlock:
 out:
  switch (ret) {
  case -EIO:
-  /* If this -EIO is due to a gpu hang, give the reset code a
-   * chance to clean up the mess. Otherwise return the proper
-   * SIGBUS. */
-  if (i915_terminally_wedged(&dev_priv->gpu_error)) {
+  /*
+   * We eat errors when the gpu is terminally wedged to avoid
+   * userspace unduly crashing (gl has no provisions for mmaps to
+   * fail). But any other -EIO isn't ours (e.g. swap in failure)
+   * and so needs to be reported.
+   */
+  if (!i915_terminally_wedged(&dev_priv->gpu_error)) {
    ret = VM_FAULT_SIGBUS;
    break;
   }
-- 
1.7.1